### PR TITLE
fixed command for install from source

### LINKED ISSUE #312: Install from source not working
The install from source is not working on the staging branch, giving ClassNotFoundException

```
(serve) ubuntu@ip-172-31-15-18:~/stage/serve$ pip install -U -e .
Obtaining file:///home/ubuntu/stage/serve
Requirement already satisfied, skipping upgrade: Pillow in /home/ubuntu/anaconda3/lib/python3.6/site-packages (from torchserve==0.0.1b20200507) (5.4.1)
Requirement already satisfied, skipping upgrade: psutil in /home/ubuntu/anaconda3/lib/python3.6/site-packages (from torchserve==0.0.1b20200507) (5.4.5)
Requirement already satisfied, skipping upgrade: future in /home/ubuntu/anaconda3/lib/python3.6/site-packages (from torchserve==0.0.1b20200507) (0.18.2)
Installing collected packages: torchserve
 Found existing installation: torchserve 0.0.1b20200507
  Uninstalling torchserve-0.0.1b20200507:
   Successfully uninstalled torchserve-0.0.1b20200507
 Running setup.py develop for torchserve
Successfully installed torchserve
(serve) ubuntu@ip-172-31-15-18:~/stage/serve$ cd ~/work
(serve) ubuntu@ip-172-31-15-18:~/work$ torchserve --ts-config config.properties
Removing orphan pid file.
(serve) ubuntu@ip-172-31-15-18:~/work$ Error: Could not find or load main class org.pytorch.serve.ModelServer
Caused by: java.lang.ClassNotFoundException: org.pytorch.serve.ModelServer
```

### PR DESCRIPTION
**Pull request by harshbafna**
## Description

This PR fixes the command for installing from source in README.

Fixes #312

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] This change requires a documentation update

## Feature/Issue validation/testing

Clone serve repo and run following command

`pip install .`

Upgrade using following command 

`pip install -U .`

Logs for successful installation and upgrade on p3.2xlarge instance using DL AMI
[build_from_source.log](https://github.com/pytorch/serve/files/4599788/build_from_source.log)

Logs for upgrading TorchServe multiple time from source
[upgrade_torchserve_from_source.log](https://github.com/pytorch/serve/files/4599790/upgrade_torchserve_from_source.log)

## Checklist:

- [x] Have you made corresponding changes to the documentation?

### CODE REVIEW

**Comment by chauhang** [comment]
@harshbafna `pip install -U .` is not working after running `pip install .` So basically if a developers needs to reinstall code from source after code changes they are unable to do so. Tested on DLAMI 27.0 p2.xlarge on a fresh machine with jdk 11 installed. Please check what got broken in the gradle units tests. Please run thorough tests for developer mode install with and without using conda. 

**Errors**

Note: the corresponding tmp log files get deleted by the time the pip install completes so unable to see actual detailed errors. 

```
        2020-05-08 11:00:43,494 [INFO ] epollEventLoopGroup-61-7 TS_METRICS - Requests5XX.Count:1|#Level:Host|#hostname:ip-172-31-31-223,timestamp:null
    
    Gradle suite > Gradle test > org.pytorch.serve.SnapshotTest.test PASSED
    
    Gradle suite > Gradle test STANDARD_OUT
        2020-05-08 11:00:43,498 [INFO ] epollEventLoopGroup-60-1 org.pytorch.serve.ModelServer - Inference model server stopped.
        2020-05-08 11:00:43,498 [INFO ] epollEventLoopGroup-60-2 org.pytorch.serve.ModelServer - Management model server stopped.
        2020-05-08 11:00:43,500 [ERROR] epollEventLoopGroup-63-1 org.pytorch.serve.TestUtils$TestHandler - Unknown exception
        io.netty.channel.unix.Errors$NativeIoException: syscall:read(..) failed: Connection reset by peer
        	at io.netty.channel.unix.FileDescriptor.readAddress(..)(Unknown Source)
    
    5 tests completed, 1 failed
    
    > Task :server:test FAILED
    > Task :server:jacocoTestReport
    
    FAILURE: Build failed with an exception.
    
    * What went wrong:
    Execution failed for task ':server:test'.
    > There were failing tests. See the report at: file:///tmp/pip-req-build-6w016uuu/frontend/server/build/reports/tests/test/index.html
    
    * Try:
    Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
    
    * Get more help at https://help.gradle.org
    
    Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
    Use '--warning-mode all' to show the individual deprecation warnings.
    See https://docs.gradle.org/6.2.2/userguide/command_line_interface.html#sec:command_line_warnings
    
    BUILD FAILED in 2m 46s
    26 actionable tasks: 26 executed
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-req-build-6w016uuu/setup.py", line 160, in <module>
        license='Apache License Version 2.0'
      File "/home/ubuntu/anaconda3/lib/python3.6/site-packages/setuptools/__init__.py", line 144, in setup
        return distutils.core.setup(**attrs)
      File "/home/ubuntu/anaconda3/lib/python3.6/distutils/core.py", line 148, in setup
        dist.run_commands()
      File "/home/ubuntu/anaconda3/lib/python3.6/distutils/dist.py", line 955, in run_commands
        self.run_command(cmd)
      File "/home/ubuntu/anaconda3/lib/python3.6/distutils/dist.py", line 974, in run_command
        cmd_obj.run()
      File "/home/ubuntu/anaconda3/lib/python3.6/site-packages/setuptools/command/install.py", line 61, in run
        return orig.install.run(self)
      File "/home/ubuntu/anaconda3/lib/python3.6/distutils/command/install.py", line 545, in run
        self.run_command('build')
      File "/home/ubuntu/anaconda3/lib/python3.6/distutils/cmd.py", line 313, in run_command
        self.distribution.run_command(command)
      File "/home/ubuntu/anaconda3/lib/python3.6/distutils/dist.py", line 974, in run_command
        cmd_obj.run()
      File "/home/ubuntu/anaconda3/lib/python3.6/distutils/command/build.py", line 135, in run
        self.run_command(cmd_name)
      File "/home/ubuntu/anaconda3/lib/python3.6/distutils/cmd.py", line 313, in run_command
        self.distribution.run_command(command)
      File "/home/ubuntu/anaconda3/lib/python3.6/distutils/dist.py", line 974, in run_command
        cmd_obj.run()
      File "/tmp/pip-req-build-6w016uuu/setup.py", line 98, in run
        self.run_command('build_frontend')
      File "/home/ubuntu/anaconda3/lib/python3.6/distutils/cmd.py", line 313, in run_command
        self.distribution.run_command(command)
      File "/home/ubuntu/anaconda3/lib/python3.6/distutils/dist.py", line 974, in run_command
        cmd_obj.run()
      File "/tmp/pip-req-build-6w016uuu/setup.py", line 85, in run
        subprocess.check_call('frontend/gradlew -p frontend clean build', shell=True)
      File "/home/ubuntu/anaconda3/lib/python3.6/subprocess.py", line 291, in check_call
        raise CalledProcessError(retcode, cmd)
    subprocess.CalledProcessError: Command 'frontend/gradlew -p frontend clean build' returned non-zero exit status 1.
    ----------------------------------------
  Rolling back uninstall of torchserve
  Moving to /home/ubuntu/anaconda3/bin/torchserve
   from /tmp/pip-uninstall-quaidv4g/torchserve
  Moving to /home/ubuntu/anaconda3/bin/torchserve-export
   from /tmp/pip-uninstall-quaidv4g/torchserve-export
  Moving to /home/ubuntu/anaconda3/lib/python3.6/site-packages/torchserve.egg-link
   from /tmp/pip-uninstall-8qtvaozr/torchserve.egg-link
ERROR: Command errored out with exit status 1: /home/ubuntu/anaconda3/bin/python -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-req-build-6w016uuu/setup.py'"'"'; __file__='"'"'/tmp/pip-req-build-6w016uuu/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-83asgq9m/install-record.txt --single-version-externally-managed --compile Check the logs for full command output.
```

**Comment by harshbafna** [comment]
@chauhang Tried following on a fresh p3.2large instance and could not reproduce the issue.

```
ubuntu@ip-172-31-40-230:~$ sudo apt-get install openjdk-11-jdk
ubuntu@ip-172-31-40-230:~$ conda install psutil future pytorch torchtext torchvision -c pytorch
ubuntu@ip-172-31-40-230:~$ git clone https://github.com/pytorch/serve
ubuntu@ip-172-31-40-230:~$ cd serve
ubuntu@ip-172-31-40-230:~/serve$ pip install .
ubuntu@ip-172-31-40-230:~/serve$ pip install -U .
```

Refer attached file for detailed logs.
[build_from_source.log](https://github.com/pytorch/serve/files/4598747/build_from_source.log)

**Comment by harshbafna** [comment]
@chauhang : Upgraded `torchserve` multiple times using `pip install -U .` on the same setup without any failures. 

Refer attached logs for more details.
[upgrade_torchserve_from_source.log](https://github.com/pytorch/serve/files/4599489/upgrade_torchserve_from_source.log)

**Comment by lokeshgupta1975** [comment]
@harshbafna Did you test this on both master and staging_0_1_1 branch?
Also, did you try to follow the installation instructions in README.md exactly? Did you try both the conda and pip installation procedures for Developer mode?

**Comment by harshbafna** [comment]
> @harshbafna Did you test this on both master and staging_0_1_1 branch?
> Also, did you try to follow the installation instructions in README.md exactly? Did you try both the conda and pip installation procedures for Developer mode?

@lokeshgupta1975 : Yes, followed the commands exactly copied from readme. The current reports are from conda environment with master branch. I will publish the the reports for staging_0_1_1 branch and python virutal envrionment in sometime.

**Comment by harshbafna** [comment]
Successfully installed and upgraded on non-conda python environment. Executed following commands :

ubuntu@ip-172-31-85-179:~$ pip install torch torchtext torchvision sentencepiece psutil future
ubuntu@ip-172-31-85-179:~$ git clone https://github.com/pytorch/serve
ubuntu@ip-172-31-85-179:~$ cd serve
ubuntu@ip-172-31-85-179:~/serve$ pip install .
ubuntu@ip-172-31-85-179:~/serve$ pip install -U .
ubuntu@ip-172-31-85-179:~/serve$ git checkout staging_0_1_1
ubuntu@ip-172-31-85-179:~/serve$ pip install -U .

Refer attached logs for more details.
[install_upgrade_from_source_non_conda.log](https://github.com/pytorch/serve/files/4600939/install_upgrade_from_source_non_conda.log)

File: README.md [hunk 1]
```diff
@@ -122,20 +137,20 @@ If you plan to develop with TorchServe and change some of the source code, you m
 1. Make your changes executable
 
     ```bash
-    pip install -e .
+    pip install .
```
Comment (aaronmarkham) [inline README.md:140]: The instruction is to make it executable, but you remove the executable flag.

File: README.md [hunk 2]
```diff
@@ -122,20 +137,20 @@ If you plan to develop with TorchServe and change some of the source code, you m
...
     ```
 
 * To develop with torch-model-archiver:
 
 ```bash
 cd serve/model-archiver
-pip install -e .
+pip install .
```
Comment (aaronmarkham) [inline README.md:147]: The instruction is to make it executable, but you remove the executable flag.

File: README.md [hunk 3]
```diff
@@ -122,20 +137,20 @@ If you plan to develop with TorchServe and change some of the source code, you m
...
 ```
 
 * To upgrade TorchServe or model archiver from source code and make changes executable, run:
 
 ```bash
-pip install -U -e .
+pip install -U .
```
Comment (aaronmarkham) [inline README.md:153]: The instruction is to make it executable, but you remove the executable flag.

File: README.md [hunk 4]
```diff
@@ -108,8 +108,23 @@ If you plan to develop with TorchServe and change some of the source code, you m
...
+
+    ```bash
+    conda install psutil future pytorch torchtext torchvision -c pytorch -c powerai
+    ```
+
+    For GPU
+
     ```bash
-    pip install psutil future
+    conda install psutil future pytorch torchtext torchvision cudatoolkit=10.1 -c pytorch -c powerai
```
Comment (aaronmarkham) [inline README.md:127]: `-c powerai` isn't really being used here...

File: README.md [hunk 5]
```diff
@@ -108,8 +108,23 @@ If you plan to develop with TorchServe and change some of the source code, you m
...
+
+    ```bash
+    pip install torch torchtext torchvision sentencepiece psutil future
+    ```
+
+* To install in Conda virtual environment
+    For CPU
+
+    ```bash
+    conda install psutil future pytorch torchtext torchvision -c pytorch -c powerai
```
Comment (aaronmarkham) [inline README.md:121]: `-c powerai` isn't really being used here...

File: README.md (hunk 1 above)
Comment (harshbafna) [reply README.md:140]: -e flag is for "editable" and not "executable", which is causing the problem when we install from source. This flag skips the gradle build and hence the frontend jar is not generated leading to "ClassNotFoundException" while starting `torchserve`

Reference : https://pip.pypa.io/en/stable/reference/pip_install/#options

File: README.md (hunk 4 above)
Comment (harshbafna) [reply README.md]: Done.

File: README.md (hunk 5 above)
Comment (harshbafna) [reply README.md]: Done.

File: README.md (hunk 3 above)
Comment (harshbafna) [reply README.md:158]: Please refer the comment above.

File: README.md (hunk 2 above)
Comment (harshbafna) [reply README.md:152]: Please refer the comment above.

File: README.md [hunk 6]
```diff
@@ -122,20 +142,20 @@ If you plan to develop with TorchServe and change some of the source code, you m
 1. Make your changes executable
```
Comment (aaronmarkham) [inline README.md:142]:
```suggestion
1. Install from source
```

File: README.md [hunk 7]
```diff
@@ -122,20 +142,20 @@ If you plan to develop with TorchServe and change some of the source code, you m
...
 
 * To develop with torch-model-archiver:
 
 ```bash
 cd serve/model-archiver
-pip install -e .
+pip install .
 ```
 
 * To upgrade TorchServe or model archiver from source code and make changes executable, run:
```
Comment (aaronmarkham) [inline README.md:155]:
```suggestion
* To upgrade TorchServe or model archiver from source code, run:
```

**Comment by aaronmarkham** [comment]
Let's close this in favor of #322 and focus efforts there. @harshbafna please review #322.

**Pull request closed without merging**